design-proposals/golden-image-delivery-and-update-pipeline.md

```diff
@@ -0,0 +1,207 @@
+
+## HCO: Integration with SSP Operator
+
+The HCO is responsible for setting VolumeImportPoller sources on the SSP operator’s CR. These volume sources are the locations used by the VolumeImportPollers to populate PVCs for the common templates.
+
+
+## OCP UI: Integration with VolumeReferences
+
+The UI currently detects if a template has a boot source by checking if the DataVolumeTemplate’s source PVC is present. The UI needs to now be capable of checking if a DataVolumeTemplate’s source VolumeReference has phase: linked as well in order to determine if a boot source is set.
+
+## Scalability
+
+The only scaling concern is that the automated process has the potential to exhaust storage resources unless a PVC garbage collection mechanism is observed by the VolumeImportPoller.
+
+## Update/Rollback Compatibility
+
+* SSP operator must make image pipelines an opt-in feature to ensure existing boot images are not overwritten. 
+* UI must be capable of detecting boot sources for DataVolumeTemplates with sources for both VolumeReferences and PVCs
+
+## Implementation Phases
+
+**Phase1** - CDI APIs
+* VolumeReference API/controller
+* VolumeImportPoller API/controller
+
+**Phase2** - SSP + Common Templates
+* Common Templates transition to VolumeReferences for DataVolumeTemplate source
+* SSP Operator creates VolumeImportPollers for automatic PVC import
+
+**Phase3** - HCO and UI
+* HCO Team integrates with SSP Operator’s API for setting volume import sources
+* UI Team transitions to detecting boot source using a linked VolumeReference rather than looking directly for a PVC of a specific name.
+
+
+
+# API Examples
+## VM + DataVolumeTemplate + VolumeReference as source
+
+```yaml
+apiVersion: cdi.kubevirt.io/v1beta1
+kind: DataVolume
+metadata:
+   name: my-fedora-vm-image
+spec:
+   pvc:
+     accessModes:
+     - ReadWriteOnce
+     resources:
+       requests:
+         storage: 2Gi
+   source:
+     volumeReference:
+       name: fedora
+       namespace: some-namespace
+```
+
+## VolumeImportPoller Example
+
+This example polls for image tags for the quay.io/kubevirt/fedora image and imports the latest image sorting tags based on semver. As new tags are introduced, the VolumeImportPoller will import them as new PVCs and automatically manage updating the corresponding VolumeReference.
+
+* **source** specifies where to poll from
+* **pollInterval** specifies how often to look for new imports 
+* **garbageCollectOutdated** specifies whether old PVCs should be cleaned up after a new PVC is imported.
+
+```yaml
+apiVersion: cdi.kubevirt.io/v1beta1
+kind: VolumeImportPoller
+metadata:
+   name: fedora-poller
+spec:
+   source:
+     registry:
+       url: docker://quay.io/kubevirt/fedora
+    pollInterval: 12h
+    garbageCollectOutdated: true
```

Comment:
We could either garbageCollect generally or specify something like keep X images, either by count or age.
Same as replica-sets stay around.
This way admins can easily roll back on images or users can hard-bind their vmis to versions if necessary (which would work anyways with garbage collection).

Reply:
> We could either garbageCollect generally or specify something like keep X images, either by count or age.

yeah, makes sense to me. Maybe we could make this a future iteration of the feature.  I'm not sure we'd need it right out of the gate. For me, It's unclear what those parameters should be at the moment (x images, by count, age, etc...). Maybe that deserves a focused discussion later on.